debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/LcrEventHandler.java

```diff
@@ -0,0 +1,139 @@
+    private final OracleOffsetContext offsetContext;
+
+    public LcrEventHandler(ErrorHandler errorHandler, EventDispatcher dispatcher, Clock clock, RelationalDatabaseSchema schema, OracleOffsetContext offsetContext) {
+        this.errorHandler = errorHandler;
+        this.dispatcher = dispatcher;
+        this.clock = clock;
+        this.schema = schema;
+        this.offsetContext = offsetContext;
+    }
+
+    @Override
+    public void processLCR(LCR lcr) throws StreamsException {
+        long scn = convertPositionToScn(lcr.getPosition());
+
+        // After a restart it may happen we get the event with the last processed SCN again
+        if (scn <= offsetContext.getScn()) {
+            LOGGER.debug("Ignoring change event with already processed SCN {}", scn);
+            return;
+        }
+
+        offsetContext.setScn(scn);
+        offsetContext.setTransactionId(lcr.getTransactionId());
+        offsetContext.setSourceTime(lcr.getSourceTime().timestampValue().toInstant());
+
+        try {
+            if(lcr instanceof RowLCR) {
+                dispatchDataChangeEvent((RowLCR) lcr);
+            }
+            else if (lcr instanceof DDLLCR) {
+                dispatchSchemaChangeEvent((DDLLCR) lcr);
+            }
+        }
+        // nothing to be done here if interrupted; the event loop will be stopped in the streaming source
+        catch (InterruptedException e) {
+            Thread.interrupted();
+            LOGGER.info("Received signal to stop, event loop will halt");
+        }
+        // XStream's receiveLCRCallback() doesn't reliably propagate exceptions, so we do that ourselves here
+        catch (Exception e) {
+            errorHandler.setProducerThrowable(e);
+        }
+    }
+
+    private long convertPositionToScn(byte[] position) {
+        try {
+            return XStreamUtility.getSCNFromPosition(position).longValue();
+        }
+        catch (SQLException | StreamsException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void dispatchDataChangeEvent(RowLCR lcr) throws InterruptedException {
+        LOGGER.debug("Processing DML event {}", lcr);
+
+        if(RowLCR.COMMIT.equals(lcr.getCommandType())) {
+            return;
+        }
+
+        TableId tableId = getTableId(lcr);
+
+        dispatcher.dispatchDataChangeEvent(
+                tableId,
+                () -> new OracleChangeRecordEmitter(offsetContext, lcr, schema.getTable(tableId), clock),
+                DataChangeEvent::new
+        );
+    }
+
+    private void dispatchSchemaChangeEvent(DDLLCR ddlLcr) throws InterruptedException {
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Processing DDL event {}", ddlLcr.getDDLText());
+        }
+
+        TableId tableId = getTableId(ddlLcr);
+
+        dispatcher.dispatchSchemaChangeEvent(
+                tableId,
+                () -> new OracleSchemaChangeEventEmitter(offsetContext, tableId, ddlLcr)
```

Comment:
What is the advantage of passing `Supplier` here instead of instance of the object?

Reply:
The idea was to delay as much work as possible until after filtering has been applied based on the given table id. Now, with the emitter approach, which represents a delay by itself, this may have become obsolete. I'll take a look.